internal/postrs/proof.go

```diff
@@ -152,7 +144,49 @@ func (v *Verifier) Close() error {
 	return nil
 }
 
-func (v *Verifier) VerifyProof(proof *shared.Proof, metadata *shared.ProofMetadata, logger *zap.Logger, cfg Config, scryptParams ScryptParams) error {
+type verifyOptions struct {
+	mode C.VerifyMode
+}
+
+type VerifyOptionFunc func(*verifyOptions)
+
+// Verify all indices in the proof.
+func VerifyAll() VerifyOptionFunc {
+	return func(o *verifyOptions) {
+		o.mode.tag = C.VerifyMode_All
+	}
+}
+
+// Verify only the selected index.
+// The `ord` is the ordinal number of the index in the proof to verify.
+func VerifyOne(ord int) VerifyOptionFunc {
+	return func(o *verifyOptions) {
+		// `o.mode` is a C tagged union
+		o.mode.tag = C.VerifyMode_One
+		o.mode.anon0 = (*(*[32]byte)(unsafe.Pointer(&C.VerifyMode_One_Body{
+			index: C.size_t(ord),
+		})))
```

Comment:
That's how C tagged enum works. The structure has a tag and a container to fit the largest variant (CGO decided to call it `anon0` (anonymous union)). The container literally contains the memory of the variant as in C. The layout is defined (https://github.com/rust-lang/rfcs/blob/master/text/2195-really-tagged-unions.md) and not "platform specific". IMHO, this isn't any more unsafe from other unsafe operations we do. It may look scary though :).

Reply:
Anyway - I split `verify_proof()` into three functions so the C tagged union is not needed on Go -> C direction anymore. However, I kept using `VerifyResult` enum in result of `verify_x()` functions.